pychiquito/dsl.py

```diff
@@ -2,122 +2,209 @@
 from enum import Enum
 from typing import Callable, Any
 from dataclasses import dataclass
+import rust_chiquito  # rust bindings
+import json
 
-from chiquito_ast import Circuit, StepType, ExposeOffset, ForwardSignal, SharedSignal
+from chiquito_ast import AST, ASTStepType, ExposeOffset
 from query import Internal, Forward, Queriable, Shared, Fixed
-from wit_gen import FixedGenContext, TraceContext
+from wit_gen import FixedGenContext, TraceContext, StepInstance
 from cb import Constraint, Typing, ToConstraint, to_constraint
+from util import CustomEncoder, F
 
 
 #######
 # dsl #
 #######
 
 
-class CircuitContext:
-    def __init__(self):
-        self.circuit = Circuit()
+class CircuitMode(Enum):
+    NoMode = 0
+    SETUP = 1
+    Trace = 2
 
-    def forward(self: CircuitContext, name: str) -> Forward:
+
+class Circuit:
+    def __init__(self: Circuit):
+        self.circuit = AST()
+        self.trace_context = TraceContext()
+        self.rust_ast_id = 0
+        self.mode = CircuitMode.SETUP
+        self.setup()
+        self.mode = CircuitMode.Trace
```

Comment:
A circuit can have many traces from many inputs, so we need another method, let's say gen_witness that is the one that sets the mode and calls self.trace and returns the witness. Everything gen_witness is called a new witness is generated

Reply:
Also, remember that the trace has an argument that is passed everytime a new witness is generated

Reply:
Both done!